src/apify/storages/base_storage.py

```diff
@@ -0,0 +1,164 @@
+                If neither `id` nor `name` are provided, the method returns the default storage associated with the actor run.
+                If the storage with the given name does not exist, it is created.
+            force_cloud (bool, optional): If set to True, it will open a storage on the Apify Platform even when running the actor locally.
+                Defaults to False.
+            config (Configuration, optional): A `Configuration` instance, uses global configuration if omitted.
+
+        Returns:
+            An instance of the storage given by `storage_class`.
+        """
+        cls._ensure_caches_initialized()
+        assert cls._cache_by_id is not None
+        assert cls._cache_by_name is not None
+
+        assert not (id and name)
+
+        used_config = config or Configuration.get_global_configuration()
+        used_client = StorageClientManager.get_storage_client(force_cloud=force_cloud)
+
+        # Fetch default name
+        if not id and not name:
+            default_id = cls._get_default_id(used_config)
+            if isinstance(used_client, ApifyClientAsync):
+                id = default_id
+            else:
+                name = default_id
+
+        # Try to get the storage instance from cache
+        cached_storage = None
+        if id:
+            cached_storage = cls._cache_by_id.get(id)
+        elif name:
+            cached_storage = cls._cache_by_name.get(name)
+
+        if cached_storage is not None:
+            # This cast is needed since MyPy doesn't understand very well that Self and Storage are the same
+            return cast(Self, cached_storage)
+
+        # Purge default storages if configured
+        if used_config.purge_on_start:
+            await _purge_default_storages(used_client)
+
+        # Create the storage
+        if id:
+            single_storage_client = cls._get_single_storage_client(id, used_client)
+            storage_info = await single_storage_client.get()
+            if not storage_info:
+                storage_label = cls._get_human_friendly_label()
+                raise RuntimeError(f'{storage_label} with id "{id}" does not exist!')
+        else:
+            storage_collection_client = cls._get_storage_collection_client(used_client)
+            storage_info = await storage_collection_client.get_or_create(name=name)
+
+        storage = cls(storage_info['id'], storage_info.get('name'), used_client)
+
+        # Cache by id and name
+        cls._cache_by_id[storage._id] = storage
+        if storage._name is not None:
+            cls._cache_by_name[storage._name] = storage
+        return storage
+
+    async def _remove_from_cache(self) -> None:
+        if self.__class__._cache_by_id is not None:
+            del self.__class__._cache_by_id[self._id]
+            if self.__class__._cache_by_name is not None:
+                if self._name is not None:
+                    del self.__class__._cache_by_name[self._name]
```

Comment:
Why so much nesting? Maybe combine the `_cache_by_name` and `_name` condition with `and`?

Reply:
Yeah, the condition was a bit different when I first wrote it, then I updated it but didn't simplify it, I will do it